daemon/logger/journald/read.go

```diff
@@ -99,101 +101,163 @@ func getAttrs(d map[string]string) []backend.LogAttr {
-		}()
+// initialSeekTail positions the journal read pointer at a journal entry
+// relative to the tail of the journal at the time of the call based on the
+// specification in r.config. It returns true if there is an entry to read at
+// the read pointer. Otherwise the read pointer is set to a conceptual position
+// which will be resolved to the desired entry (once written) by advancing
+// forward with r.j.Next() or similar.
+func (r *reader) initialSeekTail() (bool, error) {
+	var err error
+	if r.config.Until.IsZero() {
+		err = r.j.SeekTail()
+	} else {
+		err = r.j.SeekRealtime(r.config.Until)
+	}
+	if err != nil {
+		return false, err
+	}
 
-		var (
-			err          error
-			seekedToTail bool
-		)
-		if r.config.Tail >= 0 {
-			if r.config.Until.IsZero() {
-				err = r.j.SeekTail()
-				seekedToTail = true
-			} else {
-				err = r.j.SeekRealtime(r.config.Until)
-			}
-		} else {
-			if r.config.Since.IsZero() {
-				err = r.j.SeekHead()
-			} else {
-				err = r.j.SeekRealtime(r.config.Since)
+	var ok bool
+	if r.config.Tail == 0 {
+		ok, err = r.j.Previous()
+	} else {
+		var n int
+		n, err = r.j.PreviousSkip(uint(r.config.Tail))
+		ok = n > 0
+	}
+	if err != nil {
+		return ok, err
+	}
+	if !ok {
+		// The (filtered) journal has no entries. The tail is the head: all new
+		// entries which get written into the journal from this point forward
+		// should be read from the journal. However the read pointer is
+		// positioned at a conceptual position which is not condusive to reading
+		// those entries. The tail of the journal is resolved to the last entry
+		// in the journal _at the time of the first successful Previous() call_,
+		// which means that an arbitrary number of journal entries added in the
+		// interim may be skipped: race condition. While the realtime conceptual
+		// position is not so racy, it is also unhelpful: it is the timestamp
+		// past where reading should stop, so all logs that should be followed
+		// would be skipped over.
+		// Reset the read pointer position to avoid these problems.
+		return r.initialSeekHead()
+	} else if r.config.Tail == 0 {
+		// The journal read pointer is positioned at the discrete position of
+		// the journal entry _before_ the entry to send.
+		return r.j.Next()
+	}
+
+	// Check if the PreviousSkip went too far back.
+	timestamp, err := r.j.Realtime()
+	if err != nil {
+		return false, err
+	}
+	if timestamp.Before(r.config.Since) {
+		if err := r.j.SeekRealtime(r.config.Since); err != nil {
+			return false, err
+		}
+		return r.j.Next()
+	}
+	return true, nil
+}
+
+// wait blocks until the journal has new data to read, the reader's drain
+// deadline is exceeded, or the log reading consumer is gone.
+func (r *reader) wait() (bool, error) {
+	for {
+		dur := 250 * time.Millisecond
+		if !r.drainDeadline.IsZero() {
+			now := time.Now()
+			if now.Add(dur).After(r.drainDeadline) {
+				dur = r.drainDeadline.Sub(now)
+				if dur < 0 {
+					return false, nil
+				}
 			}
```

Comment:
Isn't `dur < 0` only true if `now.After(r.drainDeadline)`? Which would effectively reduce this into:

```suggestion
			if time.Now().After(r.drainDeadline) {
			    return false, nil
			}
```

making the `dur` irrelevant?

Reply:
> Isn't `dur < 0` only true if `now.After(r.drainDeadline)`?

Yes, but it is not necessarily true if `now.Add(dur).After(r.drainDeadline)`. This code is capping the wait duration to not overshoot the drain deadline, with an early exit if the deadline has already passed.

Reply:
Ugh, right! I missed the `status, err := r.j.Wait(dur)` below 🤦🏻

Reply:
Now this code block is what it should have been all along, moving but not rewriting https://github.com/moby/moby/pull/47019/commits/afb3921dcedccd84874353399289cacae6f170a8. (It's easy to make mistakes when rewriting history if you're not careful!)